plugins/Phaser/PhaserEffect.cpp

```diff
@@ -0,0 +1,613 @@
+/*
+ * PhaserEffect.cpp
+ *
+ * Copyright (c) 2019 Lost Robot <r94231@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "PhaserEffect.h"
+
+#include "embed.h"
+#include "Engine.h"
+#include "GuiApplication.h"
+#include "lmms_math.h"
+#include "MainWindow.h"
+#include "plugin_export.h"
+#include "Song.h"
+
+
+extern "C"
+{
+
+Plugin::Descriptor PLUGIN_EXPORT Phaser_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"Phaser",
+	QT_TRANSLATE_NOOP("pluginBrowser", "A versatile phaser plugin"),
+	"Lost Robot <r94231@gmail.com>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	NULL,
+	NULL
+} ;
+
+}
+
+
+constexpr float PHA_NOISE_FLOOR = 0.00001f;
+constexpr double PHA_LOG = 2.2;
```

Comment:
Please add a comment like you explained in the PR description.

Reply:
I'm not sure what you're referring to, are you talking about this?
> and "Inane", where any and all logic in the algorithm is aggressively defenestrated (see code comments for details).

If so, scroll down and you'll see a very large comment which shows all the Phaser feedback circuit modes and explains what they do.

Reply:
I mean what is `PHA_LOG` and why it's `2.2`.

Reply:
Oh, I see.  I'm actually not sure, it's just something I found in many different documentation sources on compressor attack/release coefficient calculation.  I've seen both "2.2" and "1" used.
You can find multiple papers that use the number 2.2 from a quick Google search, e.g. [this one](http://www2.ece.rochester.edu/~zduan/teaching/ece472/projects/2016/White_Dick_paper.pdf), [this one](http://www.cs.tut.fi/~sgn14006/PDF/L05-dynamics.pdf), and [this one.](https://audioordeal.co.uk/how-to-build-a-vst-compressor/)

I know that using "1" instead makes it so the attack time is how long it takes for the signal to reach `1 - 1/e` volume, so maybe 2.2 moves that over to a more meaningful value like 90% or something like that.

In any case, it's definitely a standard value, but I'm not actually entire certain of why it is chosen.

Reply:
Now I remember I saw something related on your compressor. A simple comment will help future developers to understand what it is.